File: Sorting/QuickSort/python/QuickSort.py
class QuickSort:
    def __init__(self):
        self.arr = []

    def input(self):
        self.arr = list(map(int, input('Enter the array: ').split()))

    def partition(self, start, end):
        pivot = self.arr[end]
        p_index = start
        for i in range(start, end):
            if self.arr[i] <= pivot:
                temp = self.arr[i]
                self.arr[i] = self.arr[p_index]
                self.arr[p_index] = temp
                p_index += 1
        temp = self.arr[p_index]
        self.arr[p_index] = self.arr[end]
        self.arr[end] = temp
        return p_index

    def quick_sort(self, start, end):
        if start < end:
            p_index = self.partition(start, end)
            self.quick_sort(start, p_index-1)
            self.quick_sort(p_index+1, end)

    def sort(self):
        self.quick_sort(0, len(self.arr)-1)

    def print_array(self):
        print("Array is: " + ' '.join([str(i) for i in self.arr]))


if __name__ == '__main__':
    s = QuickSort()
    s.input()
    s.print_array()
    print('Sorting...')
    s.sort()
    s.print_array()

File: Sorting/QuickSort/java/QuickSort.java
import java.util.Scanner;

public class QuickSort {
    private int[] arr;
    private int size;

    public QuickSort(int N) {
        this.size = N;
        this.arr = new int[this.size];
    }

    public void input() {
        System.out.println("Enter the array: ");
        Scanner sc = new Scanner(System.in);
        for(int i=0; i<this.size; i++)
            this.arr[i] = sc.nextInt();
        sc.close();
    }

    public int partition(int start, int end)
    {
        int pivot = this.arr[end];
        int p_index = start;
        for(int i=start; i<end; i++) {
            if(this.arr[i] <= pivot) {
                int temp = this.arr[i];
                this.arr[i] = this.arr[p_index];
                this.arr[p_index] = temp;
                p_index++;
            }
        }
        int temp = this.arr[p_index];
        this.arr[p_index] = this.arr[end];
        this.arr[end] = temp;
        return p_index;
    }

    public void quickSort(int start, int end) {
        if(start < end) {
            int p_index = partition(start, end);
            quickSort(start, p_index-1);
            quickSort(p_index+1, end);
        }
    }

    public void sort() {
        quickSort(0, this.size-1);
    }

    public void printArray() {
        System.out.print("Array is: ");
        for(int i=0; i<this.size; i++)
            System.out.print(this.arr[i] + " ");
        System.out.println();
    }
}